google-cloud-storage/src/main/java/com/google/cloud/storage/ThroughputSink.java

```diff
@@ -0,0 +1,251 @@
+      if (!(o instanceof Record)) {
+        return false;
+      }
+      Record record = (Record) o;
+      return numBytes == record.numBytes
+          && exception == record.exception
+          && Objects.equals(begin, record.begin)
+          && Objects.equals(end, record.end);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(numBytes, begin, end, exception);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("numBytes", numBytes)
+          .add("begin", begin)
+          .add("end", end)
+          .add("exception", exception)
+          .toString();
+    }
+
+    public static Record of(long numBytes, Instant begin, Instant end, boolean exception) {
+      return new Record(numBytes, begin, end, exception);
+    }
+  }
+
+  final class LoggedThroughputSink implements ThroughputSink {
+    private static final Logger LOGGER = Logger.getLogger(ThroughputSink.class.getName());
+
+    private final String prefix;
+    private final Clock clock;
+
+    private LoggedThroughputSink(String prefix, Clock clock) {
+      this.prefix = prefix;
+      this.clock = clock;
+    }
+
+    private static final double MiB = 1d / (1024 * 1024);
+
+    @Override
+    public void recordThroughput(Record r) {
+      LOGGER.info(
+          () ->
+              String.format(
+                  "{%s} (%01.03f MiB/s) %s",
+                  prefix,
+                  ((r.numBytes * MiB)
+                      / (Duration.between(r.getBegin(), r.getEnd()).toMillis() / 1000d)),
+                  r));
+    }
+
+    @Override
+    public WritableByteChannel decorate(WritableByteChannel wbc) {
+      return new ThroughputRecordingWritableByteChannel(wbc, this, clock);
+    }
+  }
+
+  final class ThroughputRecordingWritableByteChannel implements WritableByteChannel {
+    private final WritableByteChannel delegate;
+    private final ThroughputSink sink;
+    private final Clock clock;
+
+    private ThroughputRecordingWritableByteChannel(
+        WritableByteChannel delegate, ThroughputSink sink, Clock clock) {
+      this.delegate = delegate;
+      this.sink = sink;
+      this.clock = clock;
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+      boolean exception = false;
+      int remaining = src.remaining();
+      Instant begin = clock.instant();
+      try {
+        return delegate.write(src);
+      } catch (IOException e) {
+        exception = true;
+        throw e;
+      } finally {
+        Instant end = clock.instant();
+        Record record = Record.of(remaining - src.remaining(), begin, end, exception);
+        sink.recordThroughput(record);
+      }
+    }
+
+    @Override
+    public boolean isOpen() {
+      return delegate.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+      delegate.close();
+    }
+  }
+
+  final class TeeThroughputSink implements ThroughputSink {
```

Comment:
what does 'Tee' mean?

Reply:
Tee is common convention for taking an input and passing it to two different outputs.

The wikipedia article does a pretty good job giving an overview of its use in command line environments https://en.wikipedia.org/wiki/Tee_(command)